### PR TITLE
Refactor BsAlert to Glimmer component

### PR DESCRIPTION
**Pull request by simonihmig**

### CODE REVIEW

File: addon/utils/decorators/uses-transition.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  assert('You have to provide a fadeProperty for typeClass', typeof fadeProperty === 'string');
+
+  return function () {
+    return {
+      get() {
+        return !isFastBoot(this) && this.args[fadeProperty] !== false;
+      },
+    };
+  };
+}
```
Comment (jelhan) [inline addon/utils/decorators/uses-transition.js:14]: I'm not getting how this is related to a transition. It seems to guard against fastboot only. It seems to be a `@unlessFastboot` decorator.

**Review by jelhan — CHANGES_REQUESTED** [review]
Looks good to me in general. The requested change is only about the `fade` property, which doesn't seem to be used.

I noticed (again) some architectural decision, which doesn't seem to make much sense anymore to me. They don't need to be addressed while doing the refactoring. But I wanted to get the discussion rolling as we normally don't dig that deep into the code base.

It boils down to having three different states (`this.args.visible`, `this._visible` and `this.hidden`), which control the visibility of the alert (unless I missed something). That seems to be wrong to me.

File: addon/components/bs-alert.js [hunk 2]
```diff
@@ -70,23 +66,16 @@ export default class Alert extends Component {
...
+  @arg
   visible = true;
 
   /**
    * @property _visible
    * @private
    */
-  @listenTo('visible')
+  @localCopy('visible')
   _visible;
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js:77]: Yeah, I remember the controlled vs. uncontrolled discussion we had there a long time ago... I would for now concentrate on the mere technical refactoring, so we can release most stuff without breaking changes, and look for changes to the actual API later. But I think there's quite a bit that could be reconsidered...

File: addon/components/bs-alert.js [hunk 3]
```diff
@@ -55,7 +51,7 @@ export default class Alert extends Component {
    * @readonly
    * @private
    */
-  @defaultValue
+  @tracked
   hidden = !this._visible;
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js:55]: > This default value seems wrong to me. _visible is an alias for this.visible when creating an instance. So it should be either this.visible or this.args.visible. Or did I missed something?

`this.args.visible` would be wrong, as that would not take the default of `true` into account. And `this.visible` === `this._visible` and instantiation time, so does not really matter I think?

this.args.visible and this._visible are the samantically the same, except that _visible can diverge (see the other comment re controlled vs. uncontrolled), which we can revisit as a breaking API change, but not *now*. IIRC `hidden` was needed due to CSS transitions, as the outer element that has transitions defined needs to be present in DOM before the transition is triggered. But tbh I was also quite a bit confused about this when I started this refactoring...

File: addon/components/bs-alert.js [hunk 4]
```diff
@@ -95,22 +84,12 @@ export default class Alert extends Component {
    * @default true
    * @public
    */
-  @defaultValue
+  @arg
   fade = true;
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js:88]: Hm, ok. Good catch, will check and remove if you are right...

File: addon/components/bs-alert.js [hunk 5]
```diff
@@ -95,22 +84,12 @@ export default class Alert extends Component {
...
 
-  /**
-   * Computed property to set the alert class to the component div. Will be false when dismissed to have the component
-   * div (which cannot be removed form DOM by the component itself) without any markup.
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js:103]: I think actually it made things easier, as the element that receives CSS transitions is always present. But I might be confused again. Anyway, I though that changing this could be seen as a breaking change in some edge cases, so I didn't touch it yet. I remember I tried, but got failing tests, so reverted...

File: addon/components/bs-alert.js [hunk 6]
```diff
@@ -199,29 +170,24 @@ export default class Alert extends Component {
...
     } else {
-      this.set('hidden', true);
-      this.onDismissed();
+      this.hidden = true;
+      this.args.onDismissed?.();
     }
   }
 
+  @action
   _observeIsVisible() {
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js:186]: :+1:

File: addon/utils/decorators/uses-transition.js (hunk 1 above)
Comment (simonihmig) [reply addon/utils/decorators/uses-transition.js:14]: It's not really new, we had it already as a CP macro here: https://github.com/kaliber5/ember-bootstrap/blob/master/addon/utils/cp/uses-transition.js

It *does* check for `this.args.fade`, so not really just about FastBoot though!?

File: addon/components/bs-alert.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js]: done

File: addon/components/bs-alert.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/bs-alert.js]: done

File: addon/components/bs-alert.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jelhan) [reply addon/components/bs-alert.js:55]: > this.args.visible and this._visible are the samantically the same, except that _visible can diverge

Using a state, which can diverge from the passed in for a default value is confusing for me. Would it be okay for you to use `this.visible` instead?

```suggestion
  hidden = !this.visible;
```

File: addon/components/bs-alert.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jelhan) [reply addon/components/bs-alert.js:103]: I see. Maybe something we can revisit after we have discussed the controlled vs. uncontrolled state.

File: addon/utils/decorators/uses-transition.js (hunk 1 above)
Comment (jelhan) [reply addon/utils/decorators/uses-transition.js:14]: > It _does_ check for `this.args.fade`, so not really just about FastBoot though!?

Maybe I totally missed something but it doesn't seem to check `this.args.fade` but a dynamic argument passed in. The only relationship to _fade_, which I can see is the variable name `fadeProperty`.

But I agree that we can revisit this later as it was like that before.

**Review by jelhan — APPROVED** [review]
Would still prefer if we can change the default value of `hidden` to `this.visible` but I'm also okay to merge without. Your call. :smile: 

Thanks a lot for the great work!

**Comment by simonihmig** [comment]
> Would still prefer if we can change the default value of hidden to this.visible but I'm also okay to merge without. Your call.

Done. Only for you! :wink:
